src/repr/src/fixed_length.rs

```diff
@@ -0,0 +1,1005 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Borrow;
+use std::iter;
+use std::mem::size_of;
+
+use chrono::{DateTime, Utc};
+use mz_persist_types::Codec64;
+use ordered_float::OrderedFloat;
+use serde::{Deserialize, Serialize};
+use serde_big_array::BigArray;
+use timely::container::columnation::{Columnation, CopyRegion};
+use uuid::Uuid;
+
+use crate::adt::interval::Interval;
+use crate::adt::mz_acl_item::{AclItem, MzAclItem};
+use crate::adt::timestamp::CheckedTimestamp;
+use crate::row::{
+    date_to_array, naive_date_to_arrays, read_byte_array, read_date, read_naive_date, read_time,
+    time_to_arrays,
+};
+use crate::{ColumnType, Datum, Row, ScalarType, Timestamp};
+
+/// A typedef to ease use of 9-byte fixed-length representations.
+pub type Bytes9 = BytesN<9>;
+
+/// A helper trait to get references to `Row` based on type information that only manifests
+/// at runtime (typically originating from inferred schemas).
+pub trait IntoRowByTypes<'a>: Sized {
+    /// Obtains a reference to `Row` for an instance of `Self`, given a `Row` buffer and
+    /// a schema provided by `types`.
+    ///
+    /// Implementations are free to not use `row_buf` if a zero-copy implementation
+    /// can be provided. If the Row buffer is used, then the reference returned is to it.
+    /// Implementations are also free to place specific requirements on the given schema.
+    fn into_row(&'a self, row_buf: &'a mut Row, types: &[ColumnType]) -> &'a Row;
+}
```

Comment:
You don't need the lifetime on the trait, it's enough to have it on the function:

```suggestion
pub trait IntoRowByTypes {
    /// Obtains a reference to `Row` for an instance of `Self`, given a `Row` buffer and
    /// a schema provided by `types`.
    ///
    /// Implementations are free to not use `row_buf` if a zero-copy implementation
    /// can be provided. If the Row buffer is used, then the reference returned is to it.
    /// Implementations are also free to place specific requirements on the given schema.
    fn into_row<'a>(&'a self, row_buf: &'a mut Row, types: &[ColumnType]) -> &'a Row;
}
```

Reply:
Good catch, thanks!